docs/essentials/exceptions/exceptions-best-practices.md

```diff
@@ -0,0 +1,319 @@
+
+  **Programmatic checks**. The following example uses an **if** statement to check whether a connection is closed. If it isn't, the example closes the connection instead of throwing an exception.
+
+  C#
+  ```
+  if (conn.State != ConnectionState.Closed)
+  {
+      conn.Close();
+  }
+  ```
+
+  C++
+  ```
+  if (conn->State != ConnectionState::Closed)
+  {
+      conn->Close();
+  }
+  ```
+
+  Visual Basic
+  ```
+  If conn.State <> ConnectionState.Closed Then
+      conn.Close()
+  End IF
+  ```
+
+  **Exception handling**. The following example uses a **try**/**catch** block to check the connection and to throw an exception if the connection is not closed.
+
+  C#
+  ```
+  Try
+  conn.Close()
+Catch ex As InvalidOperationException
+  Console.WriteLine(ex.GetType().FullName)
+  Console.WriteLine(ex.Message)
+End Try
+  ```
+
+  C++
+  ```
+  try
+{
+  conn->Close();
+}
+catch (InvalidOperationException^ ex)
+{
+  Console::WriteLine(ex->GetType()->FullName);
+  Console::WriteLine(ex->Message);
+}
+  ```
+
+  Visual Basic
+  ```
+  Try
+  conn.Close()
+Catch ex As InvalidOperationException
+  Console.WriteLine(ex.GetType().FullName)
+  Console.WriteLine(ex.Message)
+End Try
+  ```
+
+  The method you choose depends on how often you expect the event to occur.
+
+  - Use exception handling if the event doesn't occur very often, that is, if the event is truly exceptional and indicates an error (such as an unexpected end-of-file). When you use exception handling, less code is executed in normal conditions.
+
+  - Use the programmatic method to check for errors if the event happens routinely and could be considered part of normal execution. When you check for errors programmatically, more code is executed if an exception occurs.
+
+- Use **try**/**catch** blocks around code that can potentially generate an exception, and use a **finally** block to clean up resources, if necessary. This way, the **try** statement generates the exception, the **catch** statement handles the exception, and the finally statement closes or deallocates resources whether or not an exception occurs.
+
+- In **catch** blocks, always order exceptions from the most specific to the least specific. This technique handles the specific exception before it is passed to a more general **catch** block.
+
+## Creating and raising exceptions
+
+The following list contains guidelines for creating your own exceptions and when they should be raised. For a more details, see [Design Guidelines for Exceptions](https://msdn.microsoft.com/library/ms229014).
+
+- Design classes so that an exception is never thrown in normal use. For example, a [FileStream](https://msdn.microsoft.com/library/system.io.filestream) class provides methods that help determine whether the end of the file has been reached. This avoids the exception that is thrown if you read past the end of the file. The following example shows how to read to the end of the file.
```

Comment:
We should link to the classes in the .NET Core API reference in these topics such as http://dotnet.github.io/api/System.IO.FileStream.html